From 78cc301e3c6b3f4c06e2a05b842e7696d90554ff Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:00:21 -0400
Subject: [fs] btrfs: fix warning while merging two adjacent extents

Message-id: <1407179109-13714-135-git-send-email-sandeen@redhat.com>
Patchwork-id: 87090
O-Subject: [RHEL7.1 PATCH 134/422] btrfs: fix warning while merging two adjacent extents
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Gui Hecheng <guihc.fnst@cn.fujitsu.com>

commit 3c9665df0c5d3f471b07efc32181459386678ebd
Author: Gui Hecheng <guihc.fnst@cn.fujitsu.com>
Date:   Thu Jan 23 13:41:09 2014 +0800

    btrfs: fix warning while merging two adjacent extents

    When we have two adjacent extents in relink_extent_backref,
    we try to merge them. When we use btrfs_search_slot to locate the
    slot for the current extent, we shouldn't set "ins_len = 1",
    because we will merge it into the previous extent rather than
    insert a new item. Otherwise, we may happen to create a new leaf
    in btrfs_search_slot and path->slot[0] will be 0. Then we try to
    fetch the previous item using "path->slots[0]--", and it will cause
    a warning as follows:

     [  145.713385] WARNING: CPU: 3 PID: 1796 at fs/btrfs/extent_io.c:5043 map_private_extent_buffer+0xd4/0xe0
     [  145.713387] btrfs bad mapping eb start 5337088 len 4096, wanted 167772306 8
     ...
     [  145.713462]  [<ffffffffa034b1f4>] map_private_extent_buffer+0xd4/0xe0
     [  145.713476]  [<ffffffffa030097a>] ? btrfs_free_path+0x2a/0x40
     [  145.713485]  [<ffffffffa0340864>] btrfs_get_token_64+0x64/0xf0
     [  145.713498]  [<ffffffffa033472c>] relink_extent_backref+0x41c/0x820
     [  145.713508]  [<ffffffffa0334d69>] btrfs_finish_ordered_io+0x239/0xa80

    I encounter this warning when running defrag having mkfs.btrfs
    with option -M. At the same time there are read/writes & snapshots
    running at background.

    Signed-off-by: Gui Hecheng <guihc.fnst@cn.fujitsu.com>
    Reviewed-by: Liu Bo <bo.li.liu@oracle.com>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index 44bfc1e..2bad0b8 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -2314,7 +2314,7 @@ again:
   u64 extent_len;
   struct btrfs_key found_key;
 
-  ret = btrfs_search_slot(trans, root, &key, path, 1, 1);
+  ret = btrfs_search_slot(trans, root, &key, path, 0, 1);
   if (ret < 0)
    goto out_free_path;
 
-- 
1.7.1